app/src/main/java/nl/tudelft/pixelperfect/client/ClientListener.java

```diff
@@ -28,9 +29,29 @@
      */
     public void messageReceived(Client source, Message message) {
         if (message instanceof RoleChosenMessage) {
-            RoleActivity.updateRoleAvailability((RoleChosenMessage) message);
+            handleRoleChosenMessage((RoleChosenMessage) message);
         } else if (message instanceof NewGameMessage) {
             LobbyActivity.startGame();
+        } else if (message instanceof DisconnectMessage) {
+            GameClient game = GameClient.getInstance();
+            game.disconnect();
+        }
+    }
+
+    /**
+     * Handle a received RoleChoseMessage.
+     * @param message The message to handle.
+     */
+    public void handleRoleChosenMessage(RoleChosenMessage message) {
+        System.out.println("Received role chosen message.");
+        if (message.isAllocated()) {
+            System.out.println("Server granted request for " + message.getRole().toString());
+            RoleActivity.enterLobby(message.getRole());
+        } else {
+            //RoleActivity.updateRoleAvailability((RoleChosenMessage) message);
+            //StringBuilder sb = new StringBuilder();
+            //sb.append("The role ").append(chosenRole.getRole().toString()).append(" is already taken!");
+            //RoleActivity.showMessage(sb.toString());
         }
```

Comment:
I knew you would notice this, I did this on purpose. Will either fix or remove it next PR.